pkg/monitor/controller.go

```diff
@@ -173,10 +230,50 @@ func (monitor *controllerMonitor) deleteStaleAgentCRD(old interface{}) {
 	monitor.deleteAgentCRD(node.Name)
 }
 
+func (monitor *controllerMonitor) createAgentCRDForExternalNode(obj interface{}) {
+	en := obj.(*v1alpha1.ExternalNode)
+	monitor.createAgentCRD(en.Name)
+}
+
+func (monitor *controllerMonitor) deleteAgentCRDForExternalNode(old interface{}) {
+	en, ok := old.(*v1alpha1.ExternalNode)
+	if !ok {
+		tombstone, ok := old.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			klog.Errorf("Error decoding object when deleting ExternalNode, invalid type: %v", old)
+			return
+		}
+		en, ok = tombstone.Obj.(*v1alpha1.ExternalNode)
+		if !ok {
+			klog.Errorf("Error decoding object tombstone when deleting ExternalNode, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
+	monitor.deleteAgentCRD(en.Name)
+}
+
+func (monitor *controllerMonitor) createAgentCRD(name string) {
+	klog.InfoS("Creating agent monitoring CRD", "name", name)
+	agentCRD := new(v1beta1.AntreaAgentInfo)
+	agentCRD.Name = name
+	_, err := monitor.client.CrdV1beta1().AntreaAgentInfos().Create(context.TODO(), agentCRD, metav1.CreateOptions{})
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			klog.InfoS("Skipping creating agent monitoring CRD as it already exists", "name", name)
+		} else {
+			klog.ErrorS(err, "Failed to create agent monitoring CRD")
+		}
+	}
+}
+
 func (monitor *controllerMonitor) deleteAgentCRD(name string) {
-	klog.Infof("Deleting agent monitoring CRD %s", name)
+	klog.InfoS("Deleting agent monitoring CRD", "name", name)
 	err := monitor.client.CrdV1beta1().AntreaAgentInfos().Delete(context.TODO(), name, metav1.DeleteOptions{})
 	if err != nil {
-		klog.Errorf("Failed to delete agent monitoring CRD %s: %v", name, err)
+		if errors.IsNotFound(err) {
+			klog.InfoS("Skipping deleting agent monitoring CRD as it is not found", "name", name)
+		} else {
+			klog.ErrorS(err, "Failed to delete agent monitoring CRD")
```

Comment:
We should return an error and trigger retry?

Reply:
Thanks for the comment. You are right that we should support retry in this case. To support the retry logics here, I also refactor monitor module a little by adding the corresponding queues and workers. Please help review the latest version, thanks.